cpp/src/arrow/scalar.cc

```diff
@@ -884,9 +885,24 @@ std::string Scalar::ToString() const {
     return dict_scalar->value.dictionary->ToString() + "[" +
            dict_scalar->value.index->ToString() + "]";
   }
-  auto maybe_repr = CastTo(utf8());
+
+  if (type->id() == Type::LIST || type->id() == Type::LARGE_LIST ||
+      type->id() == Type::LIST_VIEW || type->id() == Type::LARGE_LIST_VIEW ||
+      type->id() == Type::FIXED_SIZE_LIST) {
+    auto list_scalar = checked_cast<const BaseListScalar*>(this);
+    return list_scalar->value->ToString();
```

Comment:
I am currently facing a dilemma regarding the output formatting. It has become challenging to handle it in the same manner as before.

Here are the approaches I'm considering:

1. proceed with the current output format in `ToString`, then create an issue to add a feature similar to `pretty_print.cc`. However, I'm concerned about the impact on languages that are binding to the C++ code.
2. make `ToString` more extensible. How should we go about this?

I would greatly appreciate any guidance or advice on how to proceed.

cc @kou @bkietz

Reply:
Sorry. I'm missing this context. Could you explain what is the problem and what we need to solve it?

Reply:
Yes, more time has passed since we created this PR than we initially thought.

The purpose of this PR is to completely remove the `CastTo` function.

Initially, `CastTo` was being used in the Scalar's `ToString` function, and I have added code to replace it with `Cast`.

This change necessitates addressing the following issues:

- `ToString` now outputs a format different from the original, causing unit tests to fail (for example, `BaseListScalar` and `RunEndEncodedScalar`).
  - https://github.com/apache/arrow/actions/runs/7765982125/job/21181284567?pr=39192
  - Relatedly, tests in `glib` also fail.
    - https://github.com/apache/arrow/actions/runs/7765982121/job/21181284546?pr=39192
- Errors occur due to legacy `CastTo` being bound in Cython.
  - https://github.com/apache/arrow/actions/runs/7765982118/job/21181283847?pr=39192

Reply:
ToString shouldn't be extended here. I think that adding options would decrease the C++ API's usability; we rely on ToString being a simple string representation in unit testing and error messages.

Please add special cases as necessary to preserve the existing behavior for Scalar::ToString; I think it'd be worthwhile in a follow-up to remove those special cases but that'd be a breaking change and should be guarded accordingly.

Reply:
Can we use `arrow::PrettyPrint()` directly here? We can customize pretty print behavior by `arrow::PrettyPrintOptions`.